imxrt-hal/src/tempmon.rs

```diff
@@ -0,0 +1,400 @@
+//! #    imxrt_hal::ccm::perclk::CLKSEL::IPG(ipg_hz),
+//! # );
+//!
+//! // Init temperature monitor with 8Hz measure freq
+//! // 0xffff = 2 Sec. Read more at `measure_freq()`
+//! let mut temp_mon = peripherals.tempmon.init_with_measure_freq(0x1000);
+//! temp_mon.start();
+//!
+//! let mut last_temp = 0_i32;
+//! loop {
+//!     // Get the last temperature read by the measure_freq
+//!     if let Ok(temp) = temp_mon.get_temp() {
+//!         if last_temp != temp {
+//!             // temperature changed
+//!             last_temp = temp;
+//!         }
+//!         // do something else
+//!     }
+//! }
+//! ```
+//!
+//! ## Example 3
+//!
+//! Low and high temperature Interrupt
+//!
+//! enables interrupts for the selected alarm
+//!
+//! NOTE: Make sure that you implemented the TEMP_LOW_HIGH or TEMP_PANIC interrupt handler
+//!
+//! NOTE: TEMP_LOW_HIGH is triggered for `TempSensor low` and `TempSensor high`
+//!
+//! ```no_run
+//! use imxrt_hal::{self, tempmon::TempMon};
+//! use imxrt_hal::ral::interrupt;
+//!
+//! # let mut peripherals = imxrt_hal::Peripherals::take().unwrap();
+//! # let (_, ipg_hz) = peripherals.ccm.pll1.set_arm_clock(
+//! #    imxrt_hal::ccm::PLL1::ARM_HZ,
+//! #    &mut peripherals.ccm.handle,
+//! #    &mut peripherals.dcdc,
+//! # );
+//! # let mut cfg = peripherals.ccm.perclk.configure(
+//! #    &mut peripherals.ccm.handle,
+//! #    imxrt_hal::ccm::perclk::PODF::DIVIDE_3,
+//! #    imxrt_hal::ccm::perclk::CLKSEL::IPG(ipg_hz),
+//! # );
+//!
+//! // init temperature monitor with 8Hz measure freq
+//! // 0xffff = 2 Sec. Read more at `measure_freq()`
+//! let mut temp_mon = peripherals.tempmon.init_with_measure_freq(0x1000);
+//!
+//! // set low_alarm, high_alarm, and panic_alarm
+//! temp_mon.set_alarm_values(-5_000, 65_000, 95_000);
+//!
+//! // use values from registers if you like to compare it somewhere
+//! let (low_alarm, high_alarm, panic_alarm) = temp_mon.alarm_values();
+//!
+//! // enables interrupts for low_high_alarm and not for panic_alarm
+//! unsafe {
+//!     cortex_m::peripheral::NVIC::unmask(interrupt::TEMP_LOW_HIGH);
+//!     cortex_m::peripheral::NVIC::unmask(interrupt::TEMP_PANIC);
+//! }
+//!
+//! // start could fail if the module is not powered up
+//! if temp_mon.start().is_err() {
+//!     temp_mon.power_up();
+//!     temp_mon.start();
+//! }
+//!
+//! #[cortex_m_rt::interrupt]
+//! fn TEMP_LOW_HIGH() {
+//!     // disable the interrupt to avoid endless triggers
+//!     unsafe {
+//!         cortex_m::peripheral::NVIC::mask(interrupt::TEMP_LOW_HIGH);
+//!         cortex_m::peripheral::NVIC::mask(interrupt::TEMP_PANIC);
+//!     }
+//!
+//!     // don't forget to enable it after the temperature is back to normal
+//! }
+//! ```
+
+use crate::ral;
+
+/// Indicates that the temperature monitor is powered down.
+///
+/// If you receive this error, `power_up()` the temperature monitor first,
+/// and try again.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PowerDownError();
```

Comment:
An empty tuple allows anyone outside of this module to construct this error. A private field

```rust
pub struct PowerDownError(());
```

prevents users beyond this module creating this error type. Could we change this?

[With your permission](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), I'm happy to make this change myself. Otherwise, this is a note to self; I'll address before merging.

Reply:
changed. 

Sometimes I like the types in a match statement, but in this case, `Error::Other(_)` is totally fine.